jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceCompletableFuture.java

```diff
@@ -0,0 +1,131 @@
+ * available via a {@link CompletableFuture}.
+ * <p>
+ * An example usage to asynchronously read UTF-8 content is:
+ * </p>
+ * <pre>
+ *     public static class FutureUtf8String extends ContentSourceCompletableFuture&lt;String&gt;
+ *     {
+ *         Utf8StringBuilder builder = new Utf8StringBuilder();
+ *
+ *         public FutureUtf8String(Content.Source content)
+ *         {
+ *             super(content);
+ *         }
+ *
+ *         &#064;Override
+ *         protected String parse(Content.Chunk chunk) throws Throwable
+ *         {
+ *             if (chunk.hasRemaining())
+ *                 builder.append(chunk.getByteBuffer());
+ *             return chunk.isLast() ? builder.takeCompleteString(IllegalStateException::new) : null;
+ *         }
+ *     }
+ *     ...
+ *     {
+ *         new FutureUtf8String(source).thenAccept(System.err::println);
+ *     }
+ * </pre>
+ */
+public abstract class ContentSourceCompletableFuture<X> extends CompletableFuture<X>
+{
+    private final Content.Source _content;
+
+    public ContentSourceCompletableFuture(Content.Source content)
+    {
+        _content = content;
+    }
+
+    public CompletableFuture<X> parse()
+    {
+        onContentAvailable();
+        return this;
+    }
+
+    private void onContentAvailable()
+    {
+        while (true)
+        {
+            Content.Chunk chunk = _content.read();
+            if (chunk == null)
+            {
+                _content.demand(this::onContentAvailable);
+                return;
+            }
+            if (Content.Chunk.isFailure(chunk))
+            {
+                if (!chunk.isLast() && ignoreTransientFailure(chunk.getFailure()))
+                    continue;
+                completeExceptionally(chunk.getFailure());
+                return;
+            }
+
+            try
+            {
+                X x = parse(chunk);
+                if (x != null)
+                {
+                    complete(x);
+                    return;
+                }
+            }
+            catch (Throwable failure)
+            {
+                completeExceptionally(failure);
+                return;
+            }
+            finally
+            {
+                chunk.release();
+            }
+
+            if (chunk.isLast())
+            {
+                completeExceptionally(new EOFException());
+                return;
+            }
+        }
+    }
+
+    /**
+     * Called to parse a {@link org.eclipse.jetty.io.Content.Chunk}
+     * @param chunk The chunk containing content to parse. The chunk will
+     * never be a
+     * {@link org.eclipse.jetty.io.Content.Chunk#isFailure(Content.Chunk) failure chunk}.
+     * @return The parsed {@code X} instance or null if parsing is not yet complete
+     * @throws Throwable Thrown if there is an error parsing
+     */
+    protected abstract X parse(Content.Chunk chunk) throws Throwable;
```

Comment:
I think it just parses... and many parsers have an element of accumulation.  However, an implementation is free to call chunk.retain rather than accumulate.  I'll javadoc that.

Reply:
Retaining the chunks is a form of accumulation, but you're right that parsing implicitly implies some sort of retention.

Ok to keep the `parse` name, and a bit of javadoc stating this method is supposed to do some accumulation should be enough.